diff --git a/CMakeLists.txt b/CMakeLists.txt
index 84e11dff8..2a5a1cb1f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -239,6 +239,12 @@ set(WEBOTS_SAMPLE_CONTROLLER_ENV
   "PKG_CONFIG_PATH=${CMAKE_BINARY_DIR}/pkgconfig:$ENV{PKG_CONFIG_PATH}"
 )
 
+if(WIN32)
+  set(WEBOTS_BUILD_CONTROLLERS_COMMAND bash ${CMAKE_SOURCE_DIR}/scripts/build_controllers.sh)
+else()
+  set(WEBOTS_BUILD_CONTROLLERS_COMMAND ${CMAKE_SOURCE_DIR}/scripts/build_controllers.sh)
+endif()
+
 add_custom_target(webots_controller_sdk
   DEPENDS Controller CppController ode
   COMMENT "Building Webots controller SDK libraries"
@@ -248,6 +254,6 @@ add_custom_target(webots_sample_controllers
   COMMAND ${CMAKE_COMMAND} -E env
     ${WEBOTS_SAMPLE_CONTROLLER_ENV}
-    ${CMAKE_SOURCE_DIR}/scripts/build_controllers.sh
+    ${WEBOTS_BUILD_CONTROLLERS_COMMAND}
   WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
   COMMENT "Building Makefile-based controllers in projects and resources"
 )
@@ -262,8 +268,8 @@ add_custom_target(webots_sample_controllers_strict ${WEBOTS_SAMPLE_CONTROLLERS_STRICT_ALL}
     ${WEBOTS_SAMPLE_CONTROLLER_ENV}
     WEBOTS_CONTROLLERS_STRICT=1
-    ${CMAKE_SOURCE_DIR}/scripts/build_controllers.sh
+    ${WEBOTS_BUILD_CONTROLLERS_COMMAND}
   WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
   COMMENT "Building Makefile-based controllers in projects and resources with strict failure handling"
 )
