diff --git a/aten/src/ATen/native/quantized/cpu/qlinear.h b/aten/src/ATen/native/quantized/cpu/qlinear.h
index 60501c6b537..e4b148e2932 100644
--- a/aten/src/ATen/native/quantized/cpu/qlinear.h
+++ b/aten/src/ATen/native/quantized/cpu/qlinear.h
@@ -42,7 +42,11 @@ C10_API static Tensor run_pointwise_binary_tensor(
       std::string_view unary_post_op_algorithm);
 };

+#if defined(__MINGW32__)
+Tensor _weight_int4pack_mm_cpu_tensor(
+#else
 C10_API Tensor _weight_int4pack_mm_cpu_tensor(
+#endif
     const Tensor& A,
     const Tensor& B,
     const Tensor& qGroupSize,
diff --git a/c10/core/impl/LocalDispatchKeySet.cpp b/c10/core/impl/LocalDispatchKeySet.cpp
index 0b3d87fce24..a80ecedca71 100644
--- a/c10/core/impl/LocalDispatchKeySet.cpp
+++ b/c10/core/impl/LocalDispatchKeySet.cpp
@@ -13,11 +13,11 @@ namespace c10::impl {
 // logic is encapsulated in struct PODLocalDispatchKeySet.
 thread_local PODLocalDispatchKeySet raw_local_dispatch_key_set;

-#if defined(_MSC_VER) || defined(C10_ANDROID) || defined(C10_IPHONE)
+#if defined(_WIN32) || defined(C10_ANDROID) || defined(C10_IPHONE)
 LocalDispatchKeySet tls_local_dispatch_key_set() {
   return raw_local_dispatch_key_set;
 }
-#endif // defined(_MSC_VER) || defined(C10_ANDROID) || defined(C10_IPHONE)
+#endif // defined(_WIN32) || defined(C10_ANDROID) || defined(C10_IPHONE)

 void _force_tls_local_dispatch_key_set(LocalDispatchKeySet key_set) {
   raw_local_dispatch_key_set.set_included(key_set.included_);
diff --git a/c10/core/impl/LocalDispatchKeySet.h b/c10/core/impl/LocalDispatchKeySet.h
index bba089bb2ad..b9f04ea6931 100644
--- a/c10/core/impl/LocalDispatchKeySet.h
+++ b/c10/core/impl/LocalDispatchKeySet.h
@@ -62,9 +62,9 @@ struct C10_API LocalDispatchKeySet {

 // thread_local variables cannot be C10_API on Windows.
 // Inlining this seems to break AutoDispatchBelowAutograd on Android.
-#if defined(_MSC_VER) || defined(C10_ANDROID) || defined(C10_IPHONE)
+#if defined(_WIN32) || defined(C10_ANDROID) || defined(C10_IPHONE)
 C10_API LocalDispatchKeySet tls_local_dispatch_key_set();
-#else // defined(_MSC_VER) || defined(C10_ANDROID) || defined(C10_IPHONE)
+#else // defined(_WIN32) || defined(C10_ANDROID) || defined(C10_IPHONE)
 extern C10_API thread_local PODLocalDispatchKeySet raw_local_dispatch_key_set;

 inline C10_API LocalDispatchKeySet tls_local_dispatch_key_set() {
@@ -72,7 +72,7 @@ inline C10_API LocalDispatchKeySet tls_local_dispatch_key_set() {
   // because they include this header.
   return raw_local_dispatch_key_set;
 }
-#endif // defined(_MSC_VER) || defined(C10_ANDROID) || defined(C10_IPHONE)
+#endif // defined(_WIN32) || defined(C10_ANDROID) || defined(C10_IPHONE)

 // Internal, use ThreadLocalStateGuard
 C10_API void _force_tls_local_dispatch_key_set(LocalDispatchKeySet key_set);
diff --git a/c10/util/Flags.h b/c10/util/Flags.h
index 0d460db224a..ede60e8105e 100644
--- a/c10/util/Flags.h
+++ b/c10/util/Flags.h
@@ -215,7 +215,11 @@ C10_DECLARE_REGISTRY(C10FlagsRegistry, C10FlagParser, const std::string&);
   C10_DEFINE_typed_var(std::string, name, default_value, help_str)

 // DECLARE_typed_var should be used in header files and in the global namespace.
+#if defined(__MINGW32__)
+#define C10_DECLARE_typed_var(type, name) extern type FLAGS_##name
+#else
 #define C10_DECLARE_typed_var(type, name) C10_API extern type FLAGS_##name
+#endif

 #define C10_DECLARE_int(name) C10_DECLARE_typed_var(int, name)
 #define C10_DECLARE_int32(name) C10_DECLARE_int(name)
diff --git a/c10/util/env.cpp b/c10/util/env.cpp
index 673c7d1c2ce..2df3a7d2278 100644
--- a/c10/util/env.cpp
+++ b/c10/util/env.cpp
@@ -15,25 +15,28 @@ static std::shared_mutex& get_env_mutex() {
 // Set an environment variable.
 void set_env(const char* name, const char* value, bool overwrite) {
   std::lock_guard lk(get_env_mutex());
+#ifdef _WIN32
 #ifdef _MSC_VER
 #pragma warning(push)
 #pragma warning(disable : 4996)
+#endif
   if (!overwrite) {
     // NOLINTNEXTLINE(concurrency-mt-unsafe)
     if (std::getenv(name) != nullptr) {
       return;
     }
   }
-  auto full_env_variable = fmt::format("{}={}", name, value);
   // NOLINTNEXTLINE(concurrency-mt-unsafe)
-  auto err = putenv(full_env_variable.c_str());
+  auto err = _putenv_s(name, value);
   TORCH_INTERNAL_ASSERT(
       err == 0,
-      "putenv failed for environment \"",
+      "_putenv_s failed for environment \"",
       name,
       "\", the error is: ",
       err);
+#ifdef _MSC_VER
 #pragma warning(pop)
+#endif
 #else
   // NOLINTNEXTLINE(concurrency-mt-unsafe)
   auto err = setenv(name, value, static_cast<int>(overwrite));
diff --git a/c10/util/tempfile.cpp b/c10/util/tempfile.cpp
index 86a7b75f973..fcd3fcee7e2 100644
--- a/c10/util/tempfile.cpp
+++ b/c10/util/tempfile.cpp
@@ -12,6 +12,8 @@
 #include <fcntl.h>
 #include <fileapi.h>
 #include <io.h>
+#include <share.h>
+#include <sys/stat.h>
 #endif // defined(_WIN32)

 // Creates the filename pattern passed to and completed by `mkstemp`.
diff --git a/caffe2/utils/string_utils.h b/caffe2/utils/string_utils.h
index 6bf3b867b54..12bfd237836 100644
--- a/caffe2/utils/string_utils.h
+++ b/caffe2/utils/string_utils.h
@@ -1,6 +1,7 @@
 #pragma once

 #include <algorithm>
+#include <cstdint>
 #include <memory>
 #include <string>
 #include <vector>
diff --git a/caffe2/utils/threadpool/WorkersPool.h b/caffe2/utils/threadpool/WorkersPool.h
index 6504d70c25c..87bb1cc6bc9 100644
--- a/caffe2/utils/threadpool/WorkersPool.h
+++ b/caffe2/utils/threadpool/WorkersPool.h
@@ -7,9 +7,12 @@
 #include <c10/util/irange.h>
 #include <c10/util/Logging.h>

+#if defined(_WIN32)
+#include <malloc.h>
 #if defined(_MSC_VER)
 #include <intrin.h>
 #endif
+#endif

 namespace caffe2 {

@@ -36,7 +39,7 @@ struct AllocAligned {

 #if defined(__ANDROID__)
     p = memalign(kGEMMLOWPCacheLineSize, sizeof(T));
-#elif defined(_MSC_VER)
+#elif defined(_WIN32)
     p = _aligned_malloc(sizeof(T), kGEMMLOWPCacheLineSize);
 #else
     auto res = posix_memalign(&p, kGEMMLOWPCacheLineSize, sizeof(T));
@@ -54,7 +57,7 @@ struct AllocAligned {
   static void release(T* p) {
     if (p) {
       p->~T();
-#if defined(_MSC_VER)
+#if defined(_WIN32)
       _aligned_free((void*)p);
 #else
       free((void*)p);
diff --git a/cmake/Dependencies.cmake b/cmake/Dependencies.cmake
index dac0d1f41c3..013cca3d0d6 100644
--- a/cmake/Dependencies.cmake
+++ b/cmake/Dependencies.cmake
@@ -1592,10 +1592,7 @@ endif()
 # This was the default behavior before version 12.0.0.
 # Since PyTorch C API depends on it, make it available for projects that
 # depend on PyTorch.
-set(FMT_INSTALL ON)
-set(TEMP_BUILD_SHARED_LIBS ${BUILD_SHARED_LIBS})
-set(BUILD_SHARED_LIBS OFF CACHE BOOL "Build shared libs" FORCE)
-add_subdirectory(${PROJECT_SOURCE_DIR}/third_party/fmt)
+find_package(fmt CONFIG REQUIRED)

 # Disable compiler feature checks for `fmt`.
 #
@@ -1604,10 +1601,9 @@ add_subdirectory(${PROJECT_SOURCE_DIR}/third_party/fmt)
 # CMAKE_CXX_FLAGS in ways that break feature checks. Since we already know
 # `fmt` is compatible with a superset of the compilers that PyTorch is, it
 # shouldn't be too bad to just disable the checks.
-set_target_properties(fmt-header-only PROPERTIES INTERFACE_COMPILE_FEATURES "")
+set_target_properties(fmt::fmt-header-only PROPERTIES INTERFACE_COMPILE_FEATURES "")

 list(APPEND Caffe2_DEPENDENCY_LIBS fmt::fmt-header-only)
-set(BUILD_SHARED_LIBS ${TEMP_BUILD_SHARED_LIBS} CACHE BOOL "Build shared libs" FORCE)

 # ---[ Kineto
 # edge profiler depends on KinetoProfiler but it only does cpu
@@ -1695,8 +1691,9 @@ add_library(httplib INTERFACE IMPORTED)
 target_include_directories(httplib SYSTEM INTERFACE ${PROJECT_SOURCE_DIR}/third_party/cpp-httplib)

 # Include nlohmann-json
+find_package(nlohmann_json CONFIG REQUIRED)
 add_library(nlohmann INTERFACE IMPORTED)
-include_directories(nlohmann SYSTEM INTERFACE ${PROJECT_SOURCE_DIR}/third_party/nlohmann/include)
+target_link_libraries(nlohmann INTERFACE nlohmann_json::nlohmann_json)

 # Include moodycamel
 add_library(moodycamel INTERFACE IMPORTED)
diff --git a/cmake/FlatBuffers.cmake b/cmake/FlatBuffers.cmake
index 2e8e1b957db..16a4f806ba1 100644
--- a/cmake/FlatBuffers.cmake
+++ b/cmake/FlatBuffers.cmake
@@ -1,4 +1,4 @@
-set(FlatBuffers_Include ${PROJECT_SOURCE_DIR}/third_party/flatbuffers/include)
+find_path(FlatBuffers_Include flatbuffers/flatbuffers.h REQUIRED)
 file(GLOB FlatBuffers_Library_SRCS
   ${FlatBuffers_Include}/flatbuffers/*.h
 )
diff --git a/cmake/public/utils.cmake b/cmake/public/utils.cmake
index 731bbe7d218..90c26904490 100644
--- a/cmake/public/utils.cmake
+++ b/cmake/public/utils.cmake
@@ -120,7 +120,7 @@ endfunction()
 macro(caffe2_update_option variable value)
   if(CAFFE2_CMAKE_BUILDING_WITH_MAIN_REPO)
     get_property(__help_string CACHE ${variable} PROPERTY HELPSTRING)
-    set(${variable} ${value} CACHE BOOL ${__help_string} FORCE)
+    set(${variable} ${value} CACHE BOOL "${__help_string}" FORCE)
   else()
     set(${variable} ${value})
   endif()
diff --git a/setup.py b/setup.py
index 0decaf66ed6..49408949985 100644
--- a/setup.py
+++ b/setup.py
@@ -553,8 +553,28 @@ def check_submodules() -> None:
     if str2bool(os.getenv("USE_SYSTEM_LIBS")):
         return
     folders = get_submodule_folders()
+    ignored_folders = set()
+    if str2bool(os.getenv("USE_SYSTEM_BENCHMARK")):
+        ignored_folders.add(THIRD_PARTY_DIR / "benchmark")
+    if not str2bool(os.getenv("BUILD_TEST", "1")):
+        ignored_folders.add(THIRD_PARTY_DIR / "googletest")
+    if not str2bool(os.getenv("BUILD_CUSTOM_PROTOBUF", "1")):
+        ignored_folders.add(THIRD_PARTY_DIR / "protobuf")
+    if str2bool(os.getenv("USE_SYSTEM_PYBIND11")):
+        ignored_folders.add(THIRD_PARTY_DIR / "pybind11")
+    if not str2bool(os.getenv("USE_MIMALLOC", "1")):
+        ignored_folders.add(THIRD_PARTY_DIR / "mimalloc")
+    ignored_folders.update(
+        {
+            THIRD_PARTY_DIR / "flatbuffers",
+            THIRD_PARTY_DIR / "fmt",
+        }
+    )
+    folders = [folder for folder in folders if folder not in ignored_folders]
+    if str2bool(os.getenv("PYTORCH_PACKAGE_USE_DECLARED_SUBMODULES")):
+        folders = [folder for folder in folders if not not_exists_or_empty(folder)]
     # If none of the submodule folders exists, try to initialize them
-    if all(not_exists_or_empty(folder) for folder in folders):
+    elif all(not_exists_or_empty(folder) for folder in folders):
         try:
             report(" --- Trying to initialize submodules")
             start = time.time()
@@ -579,10 +599,11 @@ def check_submodules() -> None:
                 "LICENSE.txt",
             ],
         )
-    check_for_files(
-        THIRD_PARTY_DIR / "fbgemm" / "external" / "asmjit",
-        ["CMakeLists.txt"],
-    )
+    if str2bool(os.getenv("USE_FBGEMM", "1")):
+        check_for_files(
+            THIRD_PARTY_DIR / "fbgemm" / "external" / "asmjit",
+            ["CMakeLists.txt"],
+        )


 # Windows has very bad support for symbolic links.
diff --git a/torch/csrc/jit/codegen/fuser/cpu/temp_file.h b/torch/csrc/jit/codegen/fuser/cpu/temp_file.h
index dd21f7573f3..833f6251bde 100644
--- a/torch/csrc/jit/codegen/fuser/cpu/temp_file.h
+++ b/torch/csrc/jit/codegen/fuser/cpu/temp_file.h
@@ -24,7 +24,7 @@

 namespace torch::jit::fuser::cpu {

-#ifdef _MSC_VER
+#ifdef _WIN32
 inline int wmkstemps(wchar_t* tmpl, int suffix_len) {
   int len;
   wchar_t* name;
@@ -61,7 +61,7 @@ struct TempFile {
   AT_DISALLOW_COPY_AND_ASSIGN(TempFile);

   TempFile(const std::string& t, int suffix) {
-#ifdef _MSC_VER
+#ifdef _WIN32
     auto wt = c10::u8u16(t);
     std::vector<wchar_t> tt(wt.c_str(), wt.c_str() + wt.size() + 1);
     int fd = wmkstemps(tt.data(), suffix);
@@ -95,7 +95,7 @@ struct TempFile {
     AT_ASSERT(str.size() == result);
   }

-#ifdef _MSC_VER
+#ifdef _WIN32
   void close() {
     if (file_ != nullptr) {
       fclose(file_);
@@ -109,7 +109,7 @@ struct TempFile {
   }

   ~TempFile() {
-#ifdef _MSC_VER
+#ifdef _WIN32
     if (file_ != nullptr) {
       fclose(file_);
     }
diff --git a/torch/csrc/jit/frontend/strtod.cpp b/torch/csrc/jit/frontend/strtod.cpp
index daf768ee625..ae8a72d0096 100644
--- a/torch/csrc/jit/frontend/strtod.cpp
+++ b/torch/csrc/jit/frontend/strtod.cpp
@@ -26,7 +26,7 @@

 namespace torch::jit {

-#ifdef _MSC_VER
+#ifdef _WIN32
 double strtod_c(const char* nptr, char** endptr) {
   static _locale_t loc = _create_locale(LC_ALL, "C");
   return _strtod_l(nptr, endptr, loc);
diff --git a/torch/headeronly/macros/Macros.h b/torch/headeronly/macros/Macros.h
index 63aa0d20d8e..d05480c2373 100644
--- a/torch/headeronly/macros/Macros.h
+++ b/torch/headeronly/macros/Macros.h
@@ -374,7 +374,7 @@ static inline int C10_WARP_SIZE_INTERNAL() {
 #define CUDA_KERNEL_ASSERT_MSG(cond, msg)
 #define CUDA_KERNEL_ASSERT_PRINTF(cond, msg, ...)
 #define SYCL_KERNEL_ASSERT(cond)
-#elif defined(_MSC_VER)
+#elif defined(_WIN32)
 #if defined(NDEBUG)
 extern "C" {
 C10_IMPORT
diff --git a/torch/nativert/common/FileUtil.h b/torch/nativert/common/FileUtil.h
index 6fa82347ac2..da62c48d52b 100644
--- a/torch/nativert/common/FileUtil.h
+++ b/torch/nativert/common/FileUtil.h
@@ -17,7 +17,9 @@
 // This is a massive pain to have be an `int` due to the pthread implementation
 // we support, but it's far more compatible with the rest of the windows world
 // as an `int` than it would be as a `void*`
+#ifndef __MINGW32__
 using pid_t = int;
+#endif

 using uid_t = int;
 using gid_t = int;
@@ -31,8 +33,10 @@ using ssize_t = SSIZE_T;
 #define HAVE_MODE_T 1
 // The Windows headers don't define this anywhere, nor do any of the libs
 // that Folly depends on, so define it here.
+#ifndef __MINGW32__
 using mode_t = unsigned int;
 #endif
+#endif

 // Copied from folly/portability/Fcntl.h
 #define O_CLOEXEC _O_NOINHERIT
